engine/src/main/java/org/pentaho/di/job/entries/job/JobEntryJobRunner.java

```diff
@@ -0,0 +1,167 @@
+   */
+  public JobEntryJobRunner( Job job, Result result, int entryNr, LogChannelInterface log ) {
+    this.job = job;
+    this.result = result;
+    this.log = log;
+    this.entryNr = entryNr;
+    finished = false;
+  }
+
+  public void run() {
+    try {
+      if ( job.isStopped() || ( job.getParentJob() != null && job.getParentJob().isStopped() ) ) {
+        return;
+      }
+
+      // This JobEntryRunner is a replacement for the Job thread.
+      // The job thread is never started because we simply want to wait for the result.
+      //
+      ExtensionPointHandler.callExtensionPoint( log, KettleExtensionPoint.JobStart.id, getJob() );
+
+      job.fireJobStartListeners(); // Fire the start listeners
+      result = job.execute( entryNr + 1, result );
+    } catch ( KettleException e ) {
+      e.printStackTrace();
+      log.logError( "An error occurred executing this job entry : ", e );
+      result.setResult( false );
+      result.setNrErrors( 1 );
+    } finally {
+      //[PDI-14981] otherwise will get null pointer exception if 'job finished' listeners will be using it
+      job.setResult( result );
+      try {
+        ExtensionPointHandler.callExtensionPoint( log, KettleExtensionPoint.JobFinish.id, getJob() );
+        job.getJobMeta().disposeEmbeddedMetastoreProvider();
+        log.logDebug( BaseMessages.getString( PKG, "Job.Log.DisposeEmbeddedMetastore" ) );
+        job.fireJobFinishListeners();
+
+        //catch more general exception to prevent thread hanging
+      } catch ( Exception e ) {
+        result.setNrErrors( 1 );
+        result.setResult( false );
+        log.logError( BaseMessages.getString( PKG, "Job.Log.ErrorExecJob", e.getMessage() ), e );
+      }
+      job.setFinished( true );
+    }
+    finished = true;
+  }
+
+  /**
+   * @param result The result to set.
+   */
+  public void setResult( Result result ) {
+    this.result = result;
+  }
+
+  /**
+   * @return Returns the result.
+   */
+  public Result getResult() {
+    return result;
+  }
+
+  /**
+   * @return Returns the log.
+   */
+  public LogChannelInterface getLog() {
+    return log;
+  }
+
+  /**
+   * @param log The log to set.
+   */
+  public void setLog( LogChannelInterface log ) {
+    this.log = log;
+  }
+
+  /**
+   * @return Returns the job.
+   */
+  public Job getJob() {
+    return job;
+  }
+
+  /**
+   * @param job The job to set.
+   */
+  public void setJob( Job job ) {
+    this.job = job;
+  }
+
+  /**
+   * @return Returns the entryNr.
+   */
+  public int getEntryNr() {
+    return entryNr;
+  }
+
+  /**
+   * @param entryNr The entryNr to set.
+   */
+  public void setEntryNr( int entryNr ) {
+    this.entryNr = entryNr;
+  }
+
+  /**
+   * @return Returns the finished.
+   */
+  public boolean isFinished() {
+    return finished;
+  }
+
+  public void waitUntilFinished() {
+    while ( !isFinished() && !job.isStopped() ) {
+      try {
+        Thread.sleep( 0, 1 );
+      } catch ( InterruptedException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)